### PR TITLE
feat: emoji shortcut feature

### PR DESCRIPTION
**Pull request by Tony-MK**
### Feature Preview

---

Closes: https://github.com/AppFlowy-IO/AppFlowy/issues/2797

#### PR Checklist

- [X] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [X] I've listed at least one issue that this PR fixes in the description above.
- [X] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [X] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+import 'package:flowy_infra/size.dart';
+import 'package:flowy_infra_ui/style_widget/scrolling/styled_scroll_bar.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+const double emojiSizeMax = 28;
+const int emojiNumberPerRow = 8;
+const int maxEmojiCount = 40;
+
+const Color selectedItemColor = Color(0xFFE0F8FF);
```
[the remark this answers is not in the document]
Comment (Tony-MK) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart:18]: Sorry. Variable not used anyway.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+  var _selectedIndex = 0;
+
+  @override
+  void initState() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      _focusNode.requestFocus();
+      _searchEmoji();
+    });
+    super.initState();
```
[the remark this answers is not in the document]
Comment (Tony-MK) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart:57]: Do you mean `super.initState();` should be called first?

[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart:57]: Yes

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+    searchEmojiList.clear();
+
+    // Search for the emoji in the emoji category groups
+    for (int index = 0; index < widget.state.emojiCategoryGroupList.length;) {
+      searchEmojiList.addAll(
+        widget.state.emojiCategoryGroupList[index].emoji
+            .where(
+              (item) =>
+                  (query.isEmpty || item.name.toLowerCase().contains(query)) &&
```
Comment (Tony-MK) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/emoji_shortcut/emoji_shortcut_builder.dart:143]: @Xazin, Do you think the `_searchEmoji` function is okay to approve? Finding emojis without using `emoji.name.toLowerCase().contains(query)` could be improved. However, we use something similar in the `DeafultEmojiPicker` widget. Do you think I should optimize it?

**Comment by Tony-MK** [comment]
@Xazin, Can you spare some time to review this PR? If you are too busy I could ask @LucasXu0 if he has some time. Also, kindly help me with feedback if my PR is sufficient. Thank you 😄

**Comment by Xazin** [comment]
I'll have to test something before leaving a final approval, I'll see if I can get to it this weekend, sorry for the wait.

**Comment by Xazin** [comment]
@Tony-MK @LucasXu0 

The `_selectEmoji` method fails to insert the emoji and delete the search text.

```dart
// Replace text with actual emoji
await widget.editorState.apply(
  widget.editorState.transaction
    ..replaceText(
      node,
      (selection.end.offset - _emojiController.text.length - 1)
          .clamp(0, selection.end.offset),
      selection.end.offset,
      searchEmojiList[_selectedIndex].emoji,
    ),
);
```
We can't merge before this is resolved.

**Comment by Tony-MK** [comment]
Working on it...

**Comment by annieappflowy** [comment]
Pending for two months. Closing this PR for now.

**Pull request closed without merging**